### ISSUE TITLE
consecutive quickfixs produce strange result

### ISSUE THREAD

**Issue by GopherJ** [rust-analyzer/rust-analyzer#5897]
![Peek 2020-08-27 13-04](https://user-images.githubusercontent.com/33961674/91435307-6a8aa200-e866-11ea-8a5a-8a0fcbd95f4d.gif)

As you can see from here, two consecutive quickfixs produced:

```rust
use diesel::{selfuse uuid::Uuid};
```
which is wrong.

## Environments

- coc-rust-analyzer: **0.7.15**
- rust-analyzer: **e7da581**

**Comment by GopherJ** [comment]
See also : https://github.com/fannheyward/coc-rust-analyzer/issues/362

**Comment by GopherJ** [comment]
If I save current buffer after each quickfix then it seems correct, shall I consider it as a LSP's limit? 

> I must save current buffer before doing the next call

**Comment by bjorn3** [comment]
> If I save current buffer after each quickfix then it seems correct, shall I consider it as a LSP's limit?

What happens here is that you save the file. Rust-analyzer then runs `cargo check`, which gives a warning with suggested fix. Rust-analyzer forwards the fix to vim. You apply the fix. `cargo check` doesn't get rerun, as you didn't save the file. The stale fix is still visible. You apply the next fix, which causes an inconsistency as the code it applies to has moved. In vscode this isn't a problem in this case (tested it), as both fixes are on different lines. Maybe coc-rust-analyzer uses byte offsets, which go stale in this case instead of line+column indices, which are kept intact in this case?

**Comment by kjeremy** [comment]
I wonder if the document version isnt being taken into account on the client side.

**Comment by fannheyward** [comment]
![](https://user-images.githubusercontent.com/345274/92469305-32407780-f207-11ea-924f-5eaa5ecaa759.gif)

**Comment by GopherJ** [comment]
Yes this is my question: `if I must save before applying a new quickfix` because I have formatter, it's slow to wait formatter reacts

**Comment by kjeremy** [comment]
I have a feeling one of these may be true:
* The edits we return are off for the file version (possibly #4901)
* We need to check if the file versions match on the server (the client should protect against this).
* Our custom code action request isn't checking file versions.

The client is supposed to check the version but maybe we're missing that because we use a custom lsp request/reply?

**Comment by ChayimFriedman2** [comment]
This is likely a rustc fix that went out of sync due to a missing save, and we can't really do anything here.
